experimental: Gröbner Walk

ooinaruhugh opened the pull request:
This PR supercedes #3131 and provides some variants of the Gröbner Walk.
Currently, this includes
- Standard walk from Collart, Kalkbrener and Mall (1997)
- Generic walk from Fukuda (2007)
- Perturbed walk from Amrhein, Gloor and Küchlin (1997)

Further things we want to do:
- #3131 included some more variants, we're still in the process of revising those. 
- Also, the code probably needs some tidying up.
- We're currently investigating whether we can improve the performance of this implementation. 
- Tests

ooinaruhugh committed:
Move Gröbner walk to experimental

ooinaruhugh committed:
Phase out global infoLevel

ooinaruhugh committed:
Reformat docstring

ooinaruhugh committed:
Add GroebnerWalk as rudimentary experimental package

ooinaruhugh committed:
Fix identation in docstring

ooinaruhugh committed:
Fix doctest

ooinaruhugh committed:
Add problematic example that does not work (for now)

ooinaruhugh committed:
Rename verbosity_scope for this module

ooinaruhugh committed:
Change to getter

ooinaruhugh committed:
Phase out `global infoLevel`

ooinaruhugh committed:
Reduce redundant docstrings

ooinaruhugh committed:
Replace manual computation of maximum

ooinaruhugh committed:
Add verbosity scope for `groebner_walk`

ooinaruhugh committed:
Rework `initials` into `initial_form`

ooinaruhugh committed:
Simplify `standard_step`

ooinaruhugh committed:
Reduce amount of Matrix<->MonomialOrdering conversions

fpnowell committed:
added lift comment

fpnowell committed:
commented lift, interreduce and next_weight

ooinaruhugh committed:
Add some exports

ooinaruhugh committed:
Renaming

ooinaruhugh committed:
Add toy example

ooinaruhugh committed:
Resolved questions

fpnowell committed:
added comments, examples and facet_preorder

fpnowell committed:
added folder new_generic_walk

fpnowell committed:
removed old dependencies

ooinaruhugh committed:
Make this thing standalone

ooinaruhugh committed:
Change order of `groebner_walk` arguments to most likely to be set by user

ooinaruhugh committed:
Remove unused argument

ooinaruhugh committed:
Remove a `while true` loop

ooinaruhugh committed:
Import some stuff from OSCAR

ooinaruhugh committed:
Enable debug print

ooinaruhugh committed:
Cleanup generic and perturbed walks

ooinaruhugh committed:
Add README.md

ooinaruhugh committed:
Move some Standard walk procedures to own file

ooinaruhugh committed:
Rename `difference_lead_tail` to `bounding_vectors`

ooinaruhugh committed:
Don't compute lead_exp twice(?)

ooinaruhugh committed:
Move perturbed_walk

ooinaruhugh committed:
Move to `perturbed_walk.jl`

fpnowell committed:
removed new_generic_walk, added markedGB generic walk

ooinaruhugh committed:
Remove `infoLevel`

fpnowell committed:
Merge branch 'main' of github.com:ooinaruhugh/GroebnerWalk.jl

fpnowell committed:
Merge branch 'main' of github.com:ooinaruhugh/GroebnerWalk.jl

fpnowell committed:
cleaning up

Co-authored-by: Kamillo Ferry <kafe@kafe.dev>

fpnowell committed:
Clean up

fpnowell committed:
Add a docstring

fpnowell committed:
Clean up

ooinaruhugh committed:
Separate the files

ooinaruhugh committed:
Polish words

ooinaruhugh committed:
Add simple benchmark

ooinaruhugh committed:
Fix typo

ooinaruhugh committed:
Move agk4

ooinaruhugh committed:
Change benchmark

ooinaruhugh committed:
Cleanup

fpnowell committed:
clean up: commented out new "divide"

fpnowell committed:
cleaning up, made lift_generic lighter

fpnowell committed:
Merge branch 'main'

ooinaruhugh committed:
Add LICENSE and README

ooinaruhugh committed:
Clean up the benchmark code

ooinaruhugh committed:
Rename `markedGB` to `MarkedGröbnerBasis`

ooinaruhugh committed:
Cleanup

ooinaruhugh committed:
Rename `markedGB` to `MarkedGroebnerBasis`

ooinaruhugh committed:
Avoid unnecessary copies in `facet_initials`

ooinaruhugh committed:
Change order of arguments in example

ooinaruhugh committed:
Speed up generic_walk!!

ooinaruhugh committed:
Document code

ooinaruhugh committed:
Cleanup examples

ooinaruhugh committed:
Fix difference_lead_tail for good

ooinaruhugh committed:
Scale down difference_lead_tail vectors and get rid of the zero issue

ooinaruhugh committed:
Add debug output

ooinaruhugh committed:
Touch `convert_bounding_vector`

ooinaruhugh committed:
Expect vector of polynomials in final step of `groebner_walk`

ooinaruhugh committed:
Return only the gens, not an entire Gröbner basis

ooinaruhugh committed:
Rename `interreduce_walk` to `autoreduce`

ooinaruhugh committed:
Return `gens(G)`

fpnowell committed:
Add examples

ooinaruhugh committed:
Push intermediate benchmarking results

ooinaruhugh committed:
Add the benchmarks

ooinaruhugh committed:
Add some benchmarks

ooinaruhugh committed:
Move benchmarks

ooinaruhugh committed:
Move examples around

ooinaruhugh committed:
Edit README.md

ooinaruhugh committed:
Clean up examples

ooinaruhugh committed:
Fix typo

ooinaruhugh committed:
Clean up exports

ooinaruhugh committed:
Clean up benchmarks

ooinaruhugh committed:
Change `_normal_form` to use `MPolyBuildCtx`

ooinaruhugh committed:
Remove unnecessary if condition

ooinaruhugh committed:
Add docstrings

ooinaruhugh committed:
Clean up

ooinaruhugh committed:
Add DOI to README.md

ooinaruhugh committed:
Add DOI to README.md

ooinaruhugh committed:
Bump version number

ooinaruhugh committed:
Change default perturbation degree to n

ooinaruhugh committed:
Fix perturbed walk, it actually perturbs now

ooinaruhugh committed:
Change some argument types

ooinaruhugh committed:
Autoreduce returns a Groebner basis

ooinaruhugh committed:
Add some diagnostics printing to standard_walk

ooinaruhugh committed:
Add perturbed_walk to benchmarks

ooinaruhugh committed:
Add diagnostic printing to generic_walk

ooinaruhugh committed:
Add diagnostic printing to perturbed_walk

ooinaruhugh committed:
Remove redundant assignment

ooinaruhugh committed:
Rename walk.jl to common.jl and clean up

ooinaruhugh committed:
Clean up

ooinaruhugh committed:
Add comments for missing docstrings

ooinaruhugh committed:
Bump to v1.0.1

ooinaruhugh committed:
Remove redundant `exponent_vectors` function

fingolfin commented on experimental/GroebnerWalk/Project.toml:
```diff
@@ -0,0 +1,8 @@
+name = "GroebnerWalk"
+uuid = "6dc8777d-2b0c-4de5-9e3c-426c1820f6f9"
+authors = ["Kamillo Ferry <kafe@kafe.dev>", "Francesco Nowell <francesconowell@gmail.com>"]
+version = "1.0.1"
```
I am not sure we can support "sub-projects" like that.

More importantly, I think code with a license deviating from the license of OSCAR overall is a no-go (at least inside of OSCAR, of course a seperate package based on OSCAR can do anything it wants). Especially a viral one like GPL3 -- including this here would risk "tainting" the whole project with that license... Where does this license choice come from? Is this because your code is based on GPL code written and licensed as such by others and hence must be GPL; or is this your own choice?

ooinaruhugh replied:
Regarding the license, we chose GPLv3 because we thought OSCAR is licensed under GPLv3 as well.  This might as well be changed to GPLv3+, that's fine.

Also, regarding "sub-projects", when reworking the code from #3131, I tried to incorporate as many suggestions as possible from there.

That's why we chose to make this an experimental package since 
- then we could develop the code separately from OSCAR and not clutter the main OSCAR module,
- there is a sizeable set of helper functions which might or might not stay, and
- this looked like the appropriate place while we bring the code up to standard.

Ultimately, the goal is to be part of the main OSCAR package and provide functionality similar to `groebner_basis` and its variants. I'm happy to hear about suggestions on how to make this happen.

thofma replied:
The things in `experimental/` are not julia packages themselves, but just submodules living in a separate folder. These submodules are loaded like normal Oscar code in `src/`, the only difference is that they are marked in the documentation as "experimental" (and things in `src/` should not depend on things `experimental/`). They don't have a separate license or a separate Project.toml or UUID.

P.S.: This is described in https://docs.oscar-system.org/dev/Experimental/intro/ (but things might not be as clear as we had hoped, so please let us know if there is something that can be improved).

fingolfin replied:
Yes my apologies for the confusion about the license, it was ne who was  confused about that for a moment ("brain fart") :oops:. Of course OSCAR is GPL 3 *or later* (but this is not clear in the README so we should clarify it). Still a separate license file is quite unusual, as is a separate Project.toml.

ooinaruhugh replied:
> The things in `experimental/` are not julia packages themselves, but just submodules living in a separate folder. These submodules are loaded like normal Oscar code in `src/`, the only difference is that they are marked in the documentation as "experimental" (and things in `src/` should not depend on things `experimental/`). They don't have a separate license or a separate Project.toml or UUID.
> 
> P.S.: This is described in https://docs.oscar-system.org/dev/Experimental/intro/ (but things might not be as clear as we had hoped, so please let us know if there is something that can be improved).

Ah, okay. I understood it that in theory you would just take a package that is to be moved to OSCAR and just plop it down in `experimental/` (and further down the road get the submodule absorbed in OSCAR). It wasn't clear that a separate `Project.toml` is specifically discouraged.

ooinaruhugh committed:
Remove standalone Julia package files

ooinaruhugh committed:
Remove standalone Julia package files

ooinaruhugh committed:
Update docstrings

ooinaruhugh committed:
Make some iterations more readable

ooinaruhugh committed:
Remove old large tests

ooinaruhugh commented:
I talked to @afkafkafk13 recently and we decided that the progress up to now is ready for review. New features will come with a later PR, of course corrections to the current code will be added.

fpnowell committed:
added standard walk tests

lgoettgens commented on experimental/GroebnerWalk/src/GroebnerWalk.jl:
```diff
@@ -0,0 +1,22 @@
+import Oscar: weight_ordering, ZZRingElem, MonomialOrdering, ZZMatrix, IdealGens
+import Oscar.Orderings: MatrixOrdering, _support_indices
+
+export groebner_walk
```
```suggestion
import Oscar: weight_ordering, ZZRingElem, MonomialOrdering, ZZMatrix, IdealGens
import Oscar.Orderings: MatrixOrdering, _support_indices

export groebner_walk

include("markedGB.jl")
include("common.jl")

include("standard_walk.jl")
include("generic_walk.jl")
include("perturbed_walk.jl")
```
You usually want to put imports/exports before the content of the module. Otherwise, the imports are not available yet when you try to use them

ooinaruhugh committed:
Fix errors from CI

lgoettgens commented on experimental/GroebnerWalk/src/GroebnerWalk.jl:
```diff
@@ -0,0 +1,22 @@
+function __init__()
+    add_verbosity_scope(:groebner_walk)
+
+    set_verbosity_level(:groebner_walk, 0)
```
```suggestion
```
this is not needed, see e.g. https://github.com/oscar-system/Oscar.jl/blob/344710c8292b58fd42fca71c67278e1a6d20fdbf/src/Oscar.jl#L143-L179 for the place where the other scopes are introduced

lgoettgens commented on experimental/GroebnerWalk/test/runtests.jl:
```diff
@@ -0,0 +1 @@
+include("standard_walk.jl")
```
this file is redundant, as Oscar automatically runs all files from the test directory when running tests

ooinaruhugh replied:
runtests.jl is deleted

lgoettgens commented on experimental/GroebnerWalk/src/common.jl:
```diff
@@ -0,0 +1,170 @@
+end
+
+#TODO docstring
+function weight_ordering(w::Vector{ZZRingElem}, o::MonomialOrdering)
```
This function is an enhancement to Oscar itself and does not need to get put inside `experimental`.
Please put it below https://github.com/oscar-system/Oscar.jl/blob/d1075791e3baadea53b705cf2b70040e8548428e/src/Rings/orderings.jl#L1283 or, even better, just change the signature to `weight_ordering(w::Vector{<:IntegerUnion}, o::MonomialOrdering)` to make one method work for both types of inputs

lgoettgens commented on experimental/GroebnerWalk/README.md:
````diff
@@ -0,0 +1,42 @@
+Then, we can pass the ideal to `groebner_walk` to calculate the Gröbner basis.
+Here, we want a Gröbner basis with respect to the lexicographic ordering on `R`.
+```julia
+using GroebnerWalk
````
This example seems to be a bit outdated

lgoettgens commented on experimental/GroebnerWalk/benchmark/agk.jl:
```diff
@@ -0,0 +1,16 @@
+    ]
+
+    return ideal(F), o2, o1
+end
```
```suggestion
end

```
all files in Oscar should end with a newline

ooinaruhugh committed:
Update experimental/GroebnerWalk/src/GroebnerWalk.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>
(cherry picked from commit d5658bd0108e6e14a0f41ea5421f37849363a884)

ooinaruhugh committed:
Remove redundant runtests

ooinaruhugh committed:
Change parameter type for weight_ordering to `<:IntegerUnion`

ooinaruhugh committed:
Move weight_ordering

ooinaruhugh committed:
Move weight_ordering

ooinaruhugh committed:
Update experimental/GroebnerWalk/benchmark/agk.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>
(cherry picked from commit b70f9e7e9e29e908da7750701f73282807aecccd)

lgoettgens commented on experimental/GroebnerWalk/test/standard_walk.jl:
```diff
@@ -0,0 +1,23 @@
+R1, (x,y,z) = polynomial_ring(QQ, ["x", "y", "z"])
```
Could you move all of these definitiona into your `@testset`? Background to this is that `@testset` introduced a new scope, and without this change you pollute the global namespace, and thus maybe change results of tests run after your file

ooinaruhugh committed:
Move definitions into testset

lgoettgens commented on experimental/GroebnerWalk/benchmark/cyclic.jl:
```diff
@@ -0,0 +1,78 @@
+    ]
+
+    return ideal(F), lex(R), default_ordering(R)
+end
```
```suggestion
end

```
Same issue for many more of your files

lgoettgens replied:
Please check all other files as well

ooinaruhugh committed:
Update experimental/GroebnerWalk/benchmark/cyclic.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>
(cherry picked from commit 9bc4357a727f508c061f6b51719449ad587391d4)

lgoettgens commented on experimental/GroebnerWalk/src/common.jl:
```diff
@@ -0,0 +1,161 @@
+  I::MPolyIdeal, 
+  target::MonomialOrdering = lex(base_ring(I)),
+  start::MonomialOrdering = default_ordering(base_ring(I));
+  perturbation_degree = length(gens(base_ring(I))), # meaning, n=#gens(R)
```
```suggestion
  perturbation_degree = ngens(base_ring(I)), # meaning, n=#gens(R)
```

lgoettgens commented on experimental/GroebnerWalk/src/common.jl:
```diff
@@ -0,0 +1,161 @@
+with the leading terms of G with respect to the current ordering.
+This means they are in the same cone of the Groebner fan. (cf. Lemma 2.2, Collart, Kalkbrener and Mall 1997)
+"""
+same_cone(G::Oscar.IdealGens, T::MonomialOrdering) = all(leading_term.(G; ordering=T) .== leading_term.(G; ordering=ordering(G)))
```
since this returns a bool, I personally would find it more intuitive if it was named something like `is_same_cone` or `defines_same_cone`.

lgoettgens commented on experimental/GroebnerWalk/README.md:
````diff
@@ -0,0 +1,39 @@
+Here, we want a Gröbner basis with respect to the lexicographic ordering on `R`.
+```julia
+using Oscar
+using Oscar.GroebnerWalk
````
```suggestion
```
you shouldn't need this second using line as `groebner_walk` is exported (at least after my other comments below)

lgoettgens commented on experimental/GroebnerWalk/benchmark/benchmarks.jl:
```diff
@@ -0,0 +1,60 @@
+using Oscar
+using GroebnerWalk
```
this won't work anymore
(and there are many more occurrences of `using GroebnerWalk` in your PR that need to be adapted)

lgoettgens commented on experimental/GroebnerWalk/results.csv:
```diff
@@ -0,0 +1,30 @@
+newellp1-QQ,51.5992475,1529.494271334,,811.051479458
+newellp1-Fp,9.365612875,1140.743606416,,0.165012125
+simple,1.149968916,0.25867925,,2.5334e-5
+katsura6-QQ,0.813826709,35.404766583,,
```
I don't really see the benefit of having this file in Oscar. It seems more like this is your personal benchmark result that's used in some paper.

afkafkafk13 replied:
While this is in experimental, I would keep that file. It can certainly help to detect regression, because it incorporates information on standardized examples.

The question, on how to keep or drop this data later on, when this package moves to src, is not relevant for the moment.

lgoettgens commented on experimental/GroebnerWalk/src/GroebnerWalk.jl:
```diff
@@ -0,0 +1,19 @@
+    add_verbosity_scope(:groebner_walk)
+end
+
+end
```
```suggestion
end

export groebner_walk
```
If you want to have functions available for Oscar users, you need to export them twice (once inside the module, and once at the bottom of the file).
Once you do this, all `using Oscar.GroebnerWalk` should be unnecessary.

lgoettgens commented on experimental/GroebnerWalk/src/common.jl:
```diff
@@ -0,0 +1,161 @@
+    )
+
+Compute a reduced Groebner basis w.r.t. to a monomial ordering by converting it using the Groebner Walk.
+The Groebner Walk is proposed by Collart, Kalkbrener & Mall (1997).
```
Please put your references into `doc/oscar_references.bib` and cite them here using `[BIBKEY](@cite)`.
For more details, see https://docs.oscar-system.org/dev/DeveloperDocumentation/documentation/#Updating-the-bibliography

lgoettgens commented on experimental/GroebnerWalk/src/common.jl:
```diff
@@ -0,0 +1,161 @@
+One can choose a strategy of:
+- Standard Walk (:standard) computes the Walk like as presented in Cox, Little & O'Shea (2005).
+- Generic Walk (:generic) computes the Walk as presented in Fukuda, Jensen, Lauritzen & Thomas (2005).
+- Perturbed Walk (:perturbed, with p = degree of the perturbation) computes the Walk as presented in Amrhein, Gloor & Küchlin (1997).
```
this looks a bit redundant to the algorithm keyword argument explanation below. Maybe these two lists can be merged?

lgoettgens commented on experimental/GroebnerWalk/src/common.jl:
```diff
@@ -0,0 +1,161 @@
+1 -> x + y^12 - y^8 + y^4 r
+2 -> y^16
+with respect to the ordering
+lex([x, y])
```
These doctest outputs are a bit outdated. You might need to re-run them and update the printing

lgoettgens commented on experimental/GroebnerWalk/src/common.jl:
```diff
@@ -0,0 +1,161 @@
+with respect to the ordering
+lex([x, y])
+
+julia> set_verbosity_level(:groebner_walk, 1);
```
```suggestion
julia> set_verbosity_level(:groebner_walk, 1);

```
empty line missing

lgoettgens commented on experimental/GroebnerWalk/src/generic_walk.jl:
```diff
@@ -0,0 +1,245 @@
+  lead_exp = leading_exponent_vector.(Lm)
+  
+  v = map(zip(lead_exp, exponents.(G))) do (l,T)
+    Ref(l) .- T
```
what's the purpose of using `Ref` here? I unfortunately don't see the issue you are trying to work around to think about more julian or more efficient ways to do it.

ooinaruhugh replied:
`l` is a Vector and `T` is a Vector of Vector. Since I want to subtract every element from `T` from `l`, I need to fix `l` for broadcasting with `Ref`.

lgoettgens commented on experimental/GroebnerWalk/src/generic_walk.jl:
```diff
@@ -0,0 +1,245 @@
+"""
+function facet_initials(MG::MarkedGroebnerBasis, v::Vector{ZZRingElem})
+  R = base_ring(MG)
+  inwG = Vector{MPolyRingElem}()
```
```suggestion
  inwG = elem_type(R)[]
```
this initializes the vector with the concrete mpolynomial type, instead of using the abstract mpoly type (that doesn't know about it's coefficient type)

ederc replied:
This should also be changed in `autoreduce()`.

lgoettgens commented on experimental/GroebnerWalk/src/generic_walk.jl:
```diff
@@ -0,0 +1,245 @@
+"""
+function lift_generic(MG::MarkedGroebnerBasis, H::MarkedGroebnerBasis)
+  return map(1:length(H.gens)) do i
+      H[i] - normal_form(H[i], MG)
```
```suggestion
    H[i] - normal_form(H[i], MG)
```

lgoettgens commented on experimental/GroebnerWalk/src/generic_walk.jl:
```diff
@@ -0,0 +1,245 @@
+    return false
+  else
+    i = 1
+  end
```
```suggestion
  end
  i = 1
```

ooinaruhugh committed:
Add newlines to end of file

lgoettgens commented on experimental/GroebnerWalk/src/markedGB.jl:
```diff
@@ -0,0 +1,121 @@
+
+struct MarkedGroebnerBasis
+    gens::Vector{<:MPolyRingElem}
+    markings::Vector{<:MPolyRingElem}
```
do both of these vectors contain elements over the same mpoly ring? If yes, it would be very beneficial to add a type parameter to `MarkedGroebnerBasis` and then use this to make these fields of type `Vector{T}`. This helps julia very much to infer types and make the resulting code more efficient

lgoettgens commented on experimental/GroebnerWalk/src/markedGB.jl:
```diff
@@ -0,0 +1,121 @@
+    function MarkedGroebnerBasis(gens::Vector{<:MPolyRingElem}, markings::Vector{<:MPolyRingElem})
+        if length(gens) != length(markings)
+            throw(ArgumentError("Inputs are of different length"))
+        else
```
There is `@req` to make easier input sanitation (see https://docs.oscar-system.org/dev/Hecke/manual/developer/macros/#@req)

ooinaruhugh commented:
The test for our code fail because the function that we export doesn't seem to be accessible. Looking at the other experimental packages wasn't really enlightening, because there the package is not imported and functions are "just used".

lgoettgens commented on experimental/GroebnerWalk/src/markedGB.jl:
```diff
@@ -0,0 +1,121 @@
+end
+
+base_ring(G::MarkedGroebnerBasis) = parent(first(G.gens))
+length(G::MarkedGroebnerBasis) = length(gens(G))
```
```suggestion
length(G::MarkedGroebnerBasis) = ngens(G)
ngens(G::MarkedGroebnerBasis) = length(gens(G))
```
everything implementing `gens` should implement `ngens` as well

lgoettgens commented on experimental/GroebnerWalk/src/markedGB.jl:
```diff
@@ -0,0 +1,121 @@
+    end
+end
+
+base_ring(G::MarkedGroebnerBasis) = parent(first(G.gens))
```
```suggestion
base_ring(G::MarkedGroebnerBasis) = parent(first(G.gens))
```
how do you make sure that `G.gens` has at least one element?

ooinaruhugh replied:
Not at all, but we were implementing this similarly to `IdealGens`.

Also, the reason we had to do this is pretty much the need for a version where markings can be specified, instead of an ordering.

lgoettgens reviewed (commented):
some more misc comments.
I don't have any knowledge about the involved mathematics, thus these are only comments about efficient of your code and some consistency with the rest of Oscar

lgoettgens commented:
> The test for our code fail because the function that we export doesn't seem to be accessible. Looking at the other experimental packages wasn't really enlightening, because there the package is not imported and functions are "just used".

This is indeed the standard in Oscar. submodules are not expected to be used with `using Oscar.submodule`, but you should add double exports instead (see https://github.com/oscar-system/Oscar.jl/pull/3821#discussion_r1688231270)

lgoettgens commented on experimental/GroebnerWalk/benchmark/agk.jl:
```diff
@@ -0,0 +1,18 @@
+<<<<<<< HEAD
```
```suggestion
```
some obsolete merge artifact

ooinaruhugh committed:
Add Manifest.toml for reproducibility of examples and benchmarks

ederc commented on experimental/GroebnerWalk/benchmark/benchmarks.jl:
Can one remove this empty file?

ooinaruhugh replied:
When adding an empty line at the end of some files, I might have accidentally blanked some of those. I think missed this one when fixing that.

ederc commented on experimental/GroebnerWalk/benchmark/agk.jl:
Concerning this benchmark and the others below: There is `src/Rings/special_ideals.jl` in which someone already has added the `katsura-n` examples. Maybe we can add your benchmarks there and remove these different files, I assume your benchmarks are also of interest to other developers/users.

ooinaruhugh replied:
I started adding the ideals to a `special-ideals.jl` in the experimental module. I will add the others with references, as soon as this is done I will move this to `src/Rings/special_ideals.jl` (in a future PR)

ederc commented on experimental/GroebnerWalk/docs/doc.main:
We should have a short, but still correct doc information about how to call your GBwalk functionality.

ooinaruhugh committed:
Update common.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Update common.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Export groebner_walk from experimental module

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Export groebner_walk from experimental module

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Update generic_walk.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Update generic_walk.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Update generic_walk.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Update generic_walk.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Update generic_walk.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Update generic_walk.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Update markedGB.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Update markedGB.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

ooinaruhugh committed:
Preliminary version of documentation

ooinaruhugh committed:
Preliminary version of documentation

ooinaruhugh committed:
Add citations

ooinaruhugh committed:
Add citations

ooinaruhugh committed:
Add citations

ooinaruhugh committed:
Add citations

ooinaruhugh committed:
Rename is_same_groebner_cone

ooinaruhugh committed:
Rename is_same_groebner_cone

ooinaruhugh committed:
Update doctests

ooinaruhugh committed:
Update doctests

ooinaruhugh committed:
Fix docstring

ooinaruhugh committed:
Fix docstring

ooinaruhugh committed:
Update GroebnerWalk/README.md

ooinaruhugh committed:
Update GroebnerWalk/README.md

ooinaruhugh committed:
Separate a statement

ooinaruhugh committed:
Separate a statement

ooinaruhugh committed:
Make use of TropicalGeometry/groebner_lift

ooinaruhugh committed:
Make use of TropicalGeometry/groebner_lift

ooinaruhugh committed:
Add type parameter to MarkedGroebnerBasis

ooinaruhugh committed:
Add 'experimental/GroebnerWalk/' from commit '5ea1f2321bd128cadc5e6e9b0406a5f3d07d4fcf'

git-subtree-dir: experimental/GroebnerWalk
git-subtree-mainline: 184dd4ae7be0aa181a959bffb5bc1f760f1a03c1
git-subtree-split: 5ea1f2321bd128cadc5e6e9b0406a5f3d07d4fcf

ooinaruhugh committed:
Remove 'using GroebnerWalk'

ooinaruhugh committed:
Remove 'using GroebnerWalk'

ooinaruhugh committed:
Add tests for generic_walk

ooinaruhugh committed:
Add tests for generic_walk

ooinaruhugh commented:
The recent forced pushes come from me messing up the exchange between separate repository and this PR. I'm working on getting the suggested edits from here in there.

I do see the appeal of small PRs now...

ooinaruhugh committed:
Modify benchmarks

ooinaruhugh committed:
Modify benchmarks

ooinaruhugh committed:
Move benchmark polynomial systems

ooinaruhugh committed:
Move benchmark polynomial systems

ooinaruhugh committed:
Remove thesisexamples

ooinaruhugh committed:
Remove thesisexamples

ooinaruhugh committed:
Expose newell_patch

ooinaruhugh committed:
Expose newell_patch

ooinaruhugh committed:
Update README.md

ooinaruhugh committed:
Update README.md

ooinaruhugh committed:
Update example to use exported ideal

ooinaruhugh committed:
Update example to use exported ideal

ooinaruhugh committed:
Add newline at end of file

ooinaruhugh committed:
Add newline at end of file

ooinaruhugh committed:
Properly export newell_patch_with_orderings

ooinaruhugh committed:
Properly export newell_patch_with_orderings

ooinaruhugh committed:
Add documentation

ooinaruhugh committed:
Add documentation

ooinaruhugh committed:
Better input sanitation

ooinaruhugh committed:
Better input sanitation

ooinaruhugh committed:
Merge commit 'e7c1bb607d5f18cf87e81af12007d327aa840fc0' into kf/gwalk

ooinaruhugh committed:
Add missing type parameter for MarkedGroebnerBasis

ooinaruhugh committed:
Clean up benchmarks

ooinaruhugh committed:
Add documentation for a special ideal

ooinaruhugh committed:
Fix missing docstring

fingolfin commented:
We just discussed this during triage. @ederc is fine with the "big picture" (we can improve details later). @afkafkafk13 is as well happy with it, but would like to have another look before approving.

Since this is big and has been around, it indeed seems sensible to me to merged this rather sooner and fine tune it later, but I'll leave it to the experts to make the final call.

ooinaruhugh commented:
Yeah, I agree with that. As far as I can tell, I addressed most of the previous remarks and the implemented algorithms do work correctly. I would also prefer to do any additions in a future PR.
